File: algorithms/300. Longest Increasing Subsequence/solution.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # define dp[i] stores minimal num of any increasing
        # subsequence of length i+1, for cuurent num[i+1], we
        # need to find out the max length end with n, if not
        # found, then there is a new length, if found, we
        # need to update its minimal value
        dp = []
        for n in nums:
            l, r = 0, len(dp) - 1
            while l <= r:
                m = (l + r) / 2
                if dp[m] < n:
                    l = m + 1
                else:
                    r = m - 1
            if l >= len(dp):
                dp.append(n)
            else:
                dp[l] = n
        return len(dp)

File: algorithms/300. Longest Increasing Subsequence/solution.cc
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> minimals;
        for ( int n : nums ) {
            auto p = lower_bound(minimals.begin(), minimals.end(), n);
            if ( minimals.end() == p ) {
                minimals.push_back(n);
            } else {
                *p = n;
            }
        }
        return minimals.size();
    }
};
